3.12/vm.md

```diff
@@ -0,0 +1,279 @@
+It would do the same thing as ``EXTENDED_ARG``, first decoding its own three arguments,
+then decode the following instruction (which is also assumed to have three arguments), and then...
+It *switches* on that opcode and jumps into the implementation of that instruction *past* the argument decoding step.
+
+```cc
+for (;;) {
+    _Py_CODEUNIT word = *next_instr++;
+    uint8_t opcode = word.first_byte;
+    uint32 oparg1 = word.second_byte;
+    uint32 oparg2;
+    uint32 oparg3;
+    uint32 oparg4;
+    uint32 oparg5;
+
+    switch (oparg) {
+
+        case OP_3_ARGS:
+            word = *next_instr++;
+            oparg2 = oparg2 | word.first_byte;
+            oparg3 = oparg3 | word.second_byte;
+        into_op_3_args:
+            ... // Code using oparg1, oparg2 and oparg3
+            break;
+
+        case EXTENDED_ARG_3:
+            // Decode second word of EXTENDED_ARG_3
+            word = *next_instr++;
+            oparg2 = word.first_byte;
+            oparg3 = word.second_byte;
+        into_extended_arg_3:
+            // Decode following 3-arg instruction
+            // First word
+            word = *next_instr++;
+            opcode = word.first_byte;
+            oparg1 = (oparg1 << 8) | word.second_byte;
+            // Second word
+            word = *next_instr++;
+            oparg2 = (oparg2 << 8) | word.first_byte;
+            oparg3 = (oparg3 << 8) | word.second_byte;
+            // Jump *into* the desired instruction,
+            // after the instruction decoding part
+            switch (oparg) {
+                case OP_3_args:
+                    goto into_op_3_args;
+                // A case for every 3-arg opcode
+                case EXTENDED_ARG_3:
+                    goto into_extended_arg_3;
+                default:
+                    Py_FatalError("...");
+            }
+
+    }
+}
+```
+
+The inner switch only contains cases for 2-arg and 3-arg opcodes.
+There should be a similar opcode ``EXTENDED_ARG_5`` which handles 4-arg and 5-arg opcodes.
+(For 2-arg and 4-arg opcodes, we waste some time setting ``oparg3`` or ``oparg5``,
+but this should not be a big deal as long as extended arguments are rare.)
+The original ``EXTENDED_ARG`` should still be retained for 1-arg opcodes.
+The code generator can conveniently generate the necessary labels and the switch.
+
+### Example
+
+Suppose we wanted to emit a ``BINARY_OP`` instruction with opargs ``1, 2, 1000, 4``.
+We'd emit this sequence:
+
+```
+EXTENDED_ARG_3  0, 0, 3
+BINARY_OP       1, 2, 232, 4
+```
+
+And for ``BINARY_OP 100000, 2, 1000, 4`` we'd do:
+
+```
+EXTENDED_ARG_3  1, 0, 0
+EXTENDED_ARG_3  134, 0, 3
+BINARY_OP       160, 2, 232, 4
+```
+
+Note that in this case, because ``BINARY_OP`` has _four_ arguments, we need to jump into the ``BINARY_OP`` implementation after the decoding of the second word of but before the decoding of the third and final word of the instruction.
```

Comment:
```suggestion
Note that in this case, because ``BINARY_OP`` has _four_ arguments, we need to jump into the ``BINARY_OP`` implementation after the decoding of the second word but before the decoding of the third and final word of the instruction.
```